### PR TITLE
feat: migrate external idp to other types

### PR DESCRIPTION
**Pull request by stebenz**
### Definition of Ready

- [x] I am happy with the code
- [x] Short description of the feature/issue is added in the pr description
- [x] PR is linked to the corresponding user story
- [x] Acceptance criteria are met
- [ ] All open todos and follow ups are defined in a new ticket and justified
- [x] Deviations from the acceptance criteria and design are agreed with the PO and documented.
- [x] No debug or dead code
- [x] My code has no repetitions
- [x] Critical parts are tested automatically
- [x] Where possible E2E tests are implemented
- [x] Documentation/examples are up-to-date
- [x] All non-functional requirements are met
- [ ] Functionality of the acceptance criteria is checked manually on the dev system.

### CODE REVIEW

File: internal/command/instance_idp.go [hunk 1]
```diff
@@ -97,6 +97,36 @@ func (c *Commands) UpdateInstanceGenericOIDCProvider(ctx context.Context, id str
...
+	cmds, err := preparation.PrepareCommands(ctx, c.eventstore.Filter, c.prepareMigrateInstanceOIDCToGoogleProvider(instanceAgg, writeModel, provider))
+	if err != nil {
+		return nil, err
+	}
+	pushedEvents, err := c.eventstore.Push(ctx, cmds...)
+	if err != nil {
+		return nil, err
+	}
+	return pushedEventsToObjectDetails(pushedEvents), nil
+}
```
Comment (adlerhurst) [inline internal/command/instance_idp.go:128]: can't you create a private function to have less duplicated code?

File: internal/command/instance_idp_model.go [hunk 2]
```diff
@@ -113,6 +113,10 @@ func (wm *InstanceOIDCIDPWriteModel) AppendEvents(events ...eventstore.Event) {
 			wm.OIDCIDPWriteModel.AppendEvents(&e.OIDCIDPChangedEvent)
 		case *instance.IDPRemovedEvent:
 			wm.OIDCIDPWriteModel.AppendEvents(&e.RemovedEvent)
+		case *instance.OIDCIDPMigratedAzureADEvent:
+			wm.OIDCIDPWriteModel.AppendEvents(&e.OIDCIDPMigratedAzureADEvent)
+		case *instance.OIDCIDPMigratedGoogleEvent:
+			wm.OIDCIDPWriteModel.AppendEvents(&e.OIDCIDPMigratedGoogleEvent)
```
Comment (adlerhurst) [inline internal/command/instance_idp_model.go:120]:
```suggestion
			wm.OIDCIDPWriteModel.AppendEvents(&e.OIDCIDPMigratedGoogleEvent)
```

File: internal/command/org_idp.go [hunk 3]
```diff
@@ -92,6 +92,34 @@ func (c *Commands) UpdateOrgGenericOIDCProvider(ctx context.Context, resourceOwn
...
+	cmds, err := preparation.PrepareCommands(ctx, c.eventstore.Filter, c.prepareMigrateOrgOIDCToGoogleProvider(orgAgg, writeModel, provider))
+	if err != nil {
+		return nil, err
+	}
+	pushedEvents, err := c.eventstore.Push(ctx, cmds...)
+	if err != nil {
+		return nil, err
+	}
+	return pushedEventsToObjectDetails(pushedEvents), nil
+}
```
Comment (adlerhurst) [inline internal/command/org_idp.go:121]: private function for less code duplication

File: internal/command/org_idp_model.go [hunk 4]
```diff
@@ -113,6 +113,10 @@ func (wm *OrgOIDCIDPWriteModel) AppendEvents(events ...eventstore.Event) {
 			wm.OIDCIDPWriteModel.AppendEvents(&e.OIDCIDPAddedEvent)
 		case *org.OIDCIDPChangedEvent:
 			wm.OIDCIDPWriteModel.AppendEvents(&e.OIDCIDPChangedEvent)
+		case *org.OIDCIDPMigratedAzureADEvent:
+			wm.OIDCIDPWriteModel.AppendEvents(&e.OIDCIDPMigratedAzureADEvent)
+		case *org.OIDCIDPMigratedGoogleEvent:
+			wm.OIDCIDPWriteModel.AppendEvents(&e.OIDCIDPMigratedGoogleEvent)
 		case *org.IDPRemovedEvent:
 			wm.OIDCIDPWriteModel.AppendEvents(&e.RemovedEvent)
```
Comment (adlerhurst) [inline internal/command/org_idp_model.go:122]:
```suggestion
			wm.OIDCIDPWriteModel.AppendEvents(&e.RemovedEvent)
```

File: internal/repository/idp/oidc.go [hunk 5]
```diff
@@ -162,3 +162,93 @@ func OIDCIDPChangedEventMapper(event *repository.Event) (eventstore.Event, error
...
+	return nil
+}
+
+func OIDCIDPMigratedAzureADEventMapper(event *repository.Event) (eventstore.Event, error) {
+	e, err := AzureADIDPAddedEventMapper(event)
+	if err != nil {
+		return nil, err
+	}
+
+	return &OIDCIDPMigratedAzureADEvent{AzureADIDPAddedEvent: *e.(*AzureADIDPAddedEvent)}, nil
```
Comment (adlerhurst) [inline internal/repository/idp/oidc.go:210]: is this pointer dereference correct?

File: internal/repository/idp/oidc.go [hunk 6]
```diff
@@ -162,3 +162,93 @@ func OIDCIDPChangedEventMapper(event *repository.Event) (eventstore.Event, error
...
+	return nil
+}
+
+func OIDCIDPMigratedGoogleEventMapper(event *repository.Event) (eventstore.Event, error) {
+	e, err := GoogleIDPAddedEventMapper(event)
+	if err != nil {
+		return nil, err
+	}
+
+	return &OIDCIDPMigratedGoogleEvent{GoogleIDPAddedEvent: *e.(*GoogleIDPAddedEvent)}, nil
```
Comment (adlerhurst) [inline internal/repository/idp/oidc.go:253]: same here

**Review by adlerhurst — CHANGES_REQUESTED** [review]

File: internal/repository/idp/oidc.go (hunk 5 above)
Comment (stebenz) [reply internal/repository/idp/oidc.go:210]: the OIDCIDPMigratedAzureADEvent contains the AzureADIDPAddedEvent and so here uses the AzureADIDPAddedEventMapper to get a pointer of AzureADIDPAddedEvent which gets dereferenced and filled in

File: internal/repository/idp/oidc.go (hunk 6 above)
Comment (stebenz) [reply internal/repository/idp/oidc.go:253]: same here but just with google instead of azure

**Review by adlerhurst — APPROVED** [review]

**Pull request merged**